### PR TITLE
Implementation of exact algorithms for maximum independent set

### PR DESCRIPTION
**Pull request by Onchao**

### CODE REVIEW

File: include/independent_set/IndependentSet.hpp [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    std::vector<NetworKit::node> getAllNodes() const;
+    std::vector<NetworKit::node> getMirrors(NetworKit::node v) const;
+    void dfs(NetworKit::node v, std::vector<bool>& visited);
+
+    NetworKit::node getMinimumDegreeNode() const;
+    NetworKit::node getMaximumDegreeNode() const;
+    void removeElements(std::vector<NetworKit::node> nodes);
+    void restoreElements(
+        std::vector<NetworKit::node>& nodes, 
+        EdgeSet& edges);
```
Comment (krzysztof-turowski) [inline include/independent_set/IndependentSet.hpp:77]: Eventually this should fit one line

File: test/testIndependentSet.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
+#include <gtest/gtest.h>
```
Comment (krzysztof-turowski) [inline test/testIndependentSet.cpp:1]: The examples are fine - but I have a small question: why did you decide against mirroring the outline of other tests? Do you think it's any better?

Please let me know, I still haven't decided on this, and it'd be great to have some input, as gtest is a little unwieldy.

File: include/independent_set/IndependentSet.hpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+    /**
+     * Actual recursive function
+     */
+    virtual std::vector<NetworKit::node> recursive() = 0;
+};
+
+/**
+ * @ingroup independentSet
+ * The class for the mis1 algorithm.
```
Comment (krzysztof-turowski) [inline include/independent_set/IndependentSet.hpp:121]: Could you provide at least the authors of each algorithm?

File: cpp/independent_set/IndependentSet.cpp [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    });
+    std::cout << std::endl;
+    graph->forEdges([&](NetworKit::node u, NetworKit::node v) {
+        std::cout << "{" << u << "," << v << "}, ";
+    });
+    std::cout << std::endl;
+}
+
+std::vector<NetworKit::node> IndependentSet::getNeighbors(NetworKit::node v) const {
+    std::vector<NetworKit::node> neighbors;
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:30]: Please reserve the appropriate size for the vector in advance.

File: cpp/independent_set/IndependentSet.cpp [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    std::vector<NetworKit::node> neighbors2;
+    for (auto x : visitedNodes) {
+        if (!visitedOnGraph[x]) {
+            visitedOnGraph[x] = true;
+            neighbors2.push_back(x);
+        }
+    }
+    return neighbors2;
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:65]: Wouldn't it be simpler to prepare (1) set of neighbors of neighbors, (2) set of NeighborsPlus - and then subtract 2 from 1?

File: cpp/independent_set/IndependentSet.cpp [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+IndependentSet::EdgeSet IndependentSet::getAllEdges() const {
+    EdgeSet edges(edgeComparator);
+    graph->forEdges([&](NetworKit::node u, NetworKit::node v) {
+        edges.insert(NetworKit::Edge(u, v, true));
+    });
+    return edges;
+}
+
+std::vector<NetworKit::node> IndependentSet::getAllNodes() const {
+    std::vector<NetworKit::node> nodes;
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:101]: return std::vector<>(graph->nodeRange.begin(), graph->nodeRange.end())   should work all the same

File: cpp/independent_set/IndependentSet.cpp [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+void IndependentSet::dfs(NetworKit::node v, std::vector<bool>& visited) {
+    if (visited[v]) {
+        return;
+    }
+    visited[v] = true;
+    std::vector<NetworKit::node> neighbors = getNeighbors(v);
+    for (auto n : neighbors) {
+       dfs(n, visited);
+    }
+}
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:144]: Why can't you use default NetworKit DFS? There is some, and I guess you can pass a lambda there

File: cpp/independent_set/IndependentSet.cpp [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+    assert(!graph->isEmpty());
+    NetworKit::count minDegree = std::numeric_limits<NetworKit::count>::max();
+    NetworKit::node index;
+    graph->forNodes([&](NetworKit::node v) {
+        if (minDegree > graph->degree(v)) {
+            minDegree = graph->degree(v);
+            index = v;
+        }
+    });
+    return index;
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:156]: index = *std::min_element(graph->nodeRange().begin(), graph->nodeRange().end(), appropriate lambda) works identically

The same for the maximum degree

File: cpp/independent_set/IndependentSet.cpp [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+    std::vector<bool> testSet;
+    graph->forNodes([&](NetworKit::node v) {
+        testSet.push_back(false);
+        independentSet[v] = false;
+    });
+
+    int best = 0;
+    unsigned long long max = 1;
+    max <<= graph->numberOfNodes();
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:263]: unsigned long long max = 1 <<graph->numberOfNodes();

And please add assertion that it doesn't work for more than 63 vertices...

File: cpp/independent_set/IndependentSet.cpp [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+SIZE 5: 27107
+SIZE 6: 1995
+SIZE 7: 117
+SIZE 8: 8
+SIZE 9: 1
+
+real	0m2,069s
+user	0m2,069s
+sys	0m0,000s
+
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:317]: Eventually please remove this

File: cpp/independent_set/IndependentSet.cpp [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        size_t testSetSize = testSet.count();
+        if (testSetSize <= best) {
+            continue;
+        }
+
+        bool illegalEdge = false;
+        graph->forEdges([&](NetworKit::node u, NetworKit::node v) {
+            if(testSet[u] && testSet[v]) {
+                illegalEdge = true;
+                return;
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:276]: This return does not improve the efficiency, as we traverse through all the edges. If you want a speedup, you need to rewrite it as "for (auto e = graph->edgeRange().begin(); e != graph->edgeRange().end(); e++) {...}

File: cpp/independent_set/ExactRecursiveIndependentSet.cpp [hunk 12]
```diff
@@ -0,0 +1,755 @@
...
+
+#include <networkit/auxiliary/BucketPQ.hpp>
+#include <networkit/graph/GraphTools.hpp>
+
+#include <limits>
+
+namespace Koala {
+
+void RecursiveIndependentSet::run() {
+    std::vector<NetworKit::node> result = recursive();
```
Comment (krzysztof-turowski) [inline cpp/independent_set/ExactRecursiveIndependentSet.cpp:18]: In fact, it's even better to return just this vector

File: cpp/independent_set/ExactRecursiveIndependentSet.cpp [hunk 13]
```diff
@@ -0,0 +1,755 @@
...
+    }   
+    hasRun = true;
+}
+
+std::vector<NetworKit::node> Mis1IndependentSet::recursive() {
+    if (graph->isEmpty()) {
+        return {};
+    }
+    NetworKit::node v = getMinimumDegreeNode();
+    std::vector<NetworKit::node> selectOneOf = getNeighborsPlus(v);
```
Comment (krzysztof-turowski) [inline cpp/independent_set/ExactRecursiveIndependentSet.cpp:33]: In cases like this, I encourage to use "auto" - less bugs, less refactoring in the future for next developers.

File: cpp/independent_set/ExactRecursiveIndependentSet.cpp [hunk 14]
```diff
@@ -0,0 +1,755 @@
...
+
+    int selectedToSet;
+    std::vector<NetworKit::node> largestSet;
+    for (auto u : selectOneOf) {
+        std::vector<NetworKit::node> neighborsPlus = getNeighborsPlus(u);
```
Comment (krzysztof-turowski) [inline cpp/independent_set/ExactRecursiveIndependentSet.cpp:38]: for (auto u : getNeighborsPlus(v)) { ...}

File: cpp/independent_set/ExactRecursiveIndependentSet.cpp [hunk 15]
```diff
@@ -0,0 +1,755 @@
...
+
+                removeElements(delVerticesCase4);
+                setCase4 = recursive();
+                setCase4.push_back(u2);
+                setCase4.push_back(u3);
+                restoreElements(delVerticesCase4, connectedEdgesCase4);
+
+                if (setCase1.size() >= setCase2.size() && setCase1.size() >= setCase3.size() && setCase1.size() >= setCase4.size()) return setCase1;
+                if (setCase2.size() >= setCase1.size() && setCase2.size() >= setCase3.size() && setCase2.size() >= setCase4.size()) return setCase2;
+                if (setCase3.size() >= setCase1.size() && setCase3.size() >= setCase2.size() && setCase3.size() >= setCase4.size()) return setCase3;
```
Comment (krzysztof-turowski) [inline cpp/independent_set/ExactRecursiveIndependentSet.cpp:195]: Consider refactoring: create vector "solutions", put there all branches + apply std::min_element to pick the best one

File: cpp/independent_set/ExactRecursiveIndependentSet.cpp [hunk 16]
```diff
@@ -0,0 +1,755 @@
...
+            if (graph->degree(a) == 5) {
+                v = a;
+                w = b;
+            }
+            else {
+                v = b;
+                w = a;
+            }
+        }        
+    });
```
Comment (krzysztof-turowski) [inline cpp/independent_set/ExactRecursiveIndependentSet.cpp:326]: I don't understand this - to be honest, this preserves only the last edge incident to a vertex with deg 5? (properly sorted, of course)

File: cpp/independent_set/IndependentSet.cpp [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+void printGraph(const NetworKit::Graph* graph) { // TODO: its just debug
+    graph->forNodes([&](NetworKit::node v) {
+        std::cout << v << " ";
+    });
+    std::cout << std::endl;
+    graph->forEdges([&](NetworKit::node u, NetworKit::node v) {
+        std::cout << "{" << u << "," << v << "}, ";
+    });
+    std::cout << std::endl;
+}
```
Comment (krzysztof-turowski) [inline cpp/independent_set/IndependentSet.cpp:27]: Eventually please remove it

**Review by krzysztof-turowski — CHANGES_REQUESTED** [review]
Thank you very much for your good work, there are some suggestions for some refactorings, please let me know online or offline if you have any doubts.

File: test/testIndependentSet.cpp [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+REGISTER_TYPED_TEST_CASE_P(SimpleGraphs, WheelGraphW_8, UtilityGraphK_3_3, PetersenGraph, FruchtGraph, TroubleGraph_1);
+
+typedef testing::Types<
+    Koala::BruteForceIndependentSet,
+    Koala::Mis1IndependentSet,
+    Koala::Mis2IndependentSet,
+    Koala::Mis3IndependentSet,
+    Koala::Mis4IndependentSet,
+    Koala::Mis5IndependentSet
+    >Algorithms;
```
[the remark this answers is not in the document]
Comment (Onchao) [reply test/testIndependentSet.cpp]: using instead of typedef does not work

File: cpp/independent_set/ExactRecursiveIndependentSet.cpp (hunk 12 above)
Comment (Onchao) [reply cpp/independent_set/ExactRecursiveIndependentSet.cpp:18]: I decided to return set as in the DominatingSet

File: test/testIndependentSet.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (krzysztof-turowski) [reply test/testIndependentSet.cpp]: I tested this and it worked, why do you claim it doesn't?

using Algorithms = testing::Types<
    Koala::BruteForceIndependentSet,
    Koala::Mis1IndependentSet,
    Koala::Mis2IndependentSet,
    Koala::Mis3IndependentSet,
    Koala::Mis4IndependentSet,
    Koala::Mis5IndependentSet,
    Koala::MeasureAndConquerIndependentSet>;

File: test/testIndependentSet.cpp (hunk 2 above)
Comment (krzysztof-turowski) [reply test/testIndependentSet.cpp:1]: OK, let it be for the time being, I'll refactor it on the side

File: cpp/independent_set/ExactRecursiveIndependentSet.cpp (hunk 16 above)
Comment (krzysztof-turowski) [reply cpp/independent_set/ExactRecursiveIndependentSet.cpp]: Well, wouldn't it be faster to look for the first such edge?

And if so, consider using "for (auto v = graph->nodeRange().begin(); v != graph->nodeRange().end(); ++v) {...}

File: test/testIndependentSet.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (Onchao) [reply test/testIndependentSet.cpp]: It does work. My bad.

**Review by krzysztof-turowski — APPROVED** [review]

**Pull request merged**